File: src/bmsx/ecs/input_action_effect_system.ts
import { $ } from '../core/engine_core';
import type { PlayerInput } from '../input/playerinput';
import { ECSystem, TickGroup } from './ecsystem';
import type { WorldObject } from '../core/object/worldobject';
import { InputIntentComponent, type InputIntentBinding, type InputIntentEdgeAssignment } from '../component/inputintentcomponent';
import { ActionEffectComponent } from '../component/actioneffectcomponent';
import { InputActionEffectComponent } from '../component/inputactioneffectcomponent';
import {
	compileProgram,
	validateProgramEffects,
	type CompiledProgram,
	type CompiledBinding,
	type BindingExecutionEnv,
	type PatternPredicate,
	type EffectExecutor,
} from '../action_effects/input_action_effect_compiler';
import { isInputActionEffectProgram, type InputActionEffectProgram } from '../action_effects/input_action_effect_dsl';
import { filter_iterable } from '../utils/filter_iterable';
import { deep_clone } from '../utils/deep_clone';
import { tokenKeyFromAsset, tokenKeyFromId } from '../util/asset_tokens';

type IntentEdge = 'press' | 'hold' | 'release';

let assetProgramsValidated = false;

function validatePrimaryAssetsOnBoot(): void {
	if (assetProgramsValidated) return;
	const data = $.assets.data;
	const entries = $.asset_source.list('data');
	for (let i = 0; i < entries.length; i++) {
		const entry = entries[i]!;
		const key = tokenKeyFromAsset(entry);
		const value = (data as Record<string, unknown>)[key];
		if (!isInputActionEffectProgram(value)) continue;
		validateProgramEffects(value, entry.resid);
	}
	assetProgramsValidated = true;
}

export class InputActionEffectSystem extends ECSystem {
	private readonly compiledById = new Map<string, CompiledProgram>();
	private readonly inlineCompiled = new WeakMap<InputActionEffectProgram, CompiledProgram>();
	private readonly validatedInlinePrograms = new WeakSet<InputActionEffectProgram>();
	private readonly resolvedPrograms = new Map<string, InputActionEffectProgram>();
	private readonly missingProgramIds = new Set<string>();
	private readonly patternCache = new Map<string, PatternPredicate>();
	private readonly patternCacheMax = 256;
	private readonly customMatchScratch: boolean[] = [];
	private readonly bindingLatch = new Map<string, boolean>();
	private readonly frameLatchTouched = new Set<string>();

	constructor(priority = 0) {
		super(TickGroup.Input, priority);
		validatePrimaryAssetsOnBoot();
		this.__ecsId = 'inputActionEffectSystem';
	}

	public override update(): void {
		this.frameLatchTouched.clear();
		this.processInputIntents();
		// InputActionEffect programs run early (TickGroup.Input) so FSM/input checks see consumed events and effect triggers immediately.
		this.processInputActionPrograms();
		const latchedKeys = Array.from(this.bindingLatch.keys());
		for (let idx = 0; idx < latchedKeys.length; idx++) {
			const key = latchedKeys[idx]!;
			if (!this.frameLatchTouched.has(key)) this.bindingLatch.delete(key);
		}
	}

	private processInputIntents(): void {
		const world = $.world;
		for (const [owner, component] of filter_iterable(world.objects_with_components(InputIntentComponent, { scope: 'active' }), ([obj]) => this.isEligibleObject(obj))) {
			if (!component.bindings || component.bindings.length === 0) continue;
			const input = this.resolveIntentPlayerInput(component, owner);
			if (!input) continue;
			for (let index = 0; index < component.bindings.length; index++) {
				const binding = component.bindings[index]!;
				this.evaluateIntentBinding(owner, input, binding);
			}
		}
	}

	private processInputActionPrograms(): void {
		for (let [obj, component] of filter_iterable($.world.objects_with_components(InputActionEffectComponent, { scope: 'active' }), (item: [ WorldObject, InputActionEffectComponent]) => this.isEligibleObject(item[0]))) {
			const program = this.resolveCompiledProgram(component);
			const programKey = this.resolveProgramKey(component, obj);
			const componentId = component.id;

			const playerIndex = obj.player_index;
			const input = $.input.getPlayerInput(playerIndex);

			const effects = obj.get_unique_component(ActionEffectComponent);
			// Programs that emit effects require ActionEffectComponent on the same object; we fail fast to keep wiring predictable.
			if (!effects && program.usesEffectTriggers) {
				throw new Error(`[InputActionEffectSystem] Program '${programKey}' triggers effects but object '${obj.id}' (component '${componentId}') has no ActionEffectComponent.`);
			}

			const ownerId = effects ? (effects.parent.id ?? obj.id) : obj.id;

			const env: BindingExecutionEnv = {
				owner: obj,
				ownerId,
				playerIndex,
				input,
				effects,
				queuedCommands: [],
				queuedEvents: [],
			};

			this.evaluateProgram(program, env, programKey);
			const queuedCommands = env.queuedCommands;
			for (let idx = 0; idx < queuedCommands.length; idx++) {
				const cmd = queuedCommands[idx]!;
				obj.dispatch_command(cmd.event, cmd.payload);
			}
			const queuedEvents = env.queuedEvents;
			for (let idx = 0; idx < queuedEvents.length; idx++) {
				const evt = queuedEvents[idx]!;
				obj.emit_gameplay_fact(evt);
			}
		}
	}

	private evaluateIntentBinding(owner: WorldObject, input: PlayerInput, binding: InputIntentBinding): void {
		const action = binding.action?.trim();
		if (!action) {
			return;
		}
		const state = input.getActionState(action);
		if (state.justpressed && binding.press) {
			this.runIntentAssignments(owner, input, binding, 'press', binding.press);
		}
		if (state.pressed && binding.hold) {
			this.runIntentAssignments(owner, input, binding, 'hold', binding.hold);
		}
		if (state.justreleased && binding.release) {
			this.runIntentAssignments(owner, input, binding, 'release', binding.release);
		}
	}

	private runIntentAssignments(
		owner: WorldObject,
		input: PlayerInput,
		binding: InputIntentBinding,
		edge: IntentEdge,
		spec: InputIntentEdgeAssignment,
	): void {
		const assignments = Array.isArray(spec) ? spec : [spec];
		for (let i = 0; i < assignments.length; i++) {
			const assignment = assignments[i];
			if (!assignment) continue;
			const path = assignment.path?.trim();
			if (!path) continue;
			const shouldClear = assignment.clear === true || (assignment.value === undefined && edge === 'release');
			const resolvedValue = shouldClear
				? undefined
				: assignment.value === undefined
					? edge === 'hold' || edge === 'press'
						? true
						: undefined
					: assignment.value;
			this.assignOwnerPath(owner, path, resolvedValue, shouldClear);
			if (assignment.consume === true) {
				input.consumeAction(binding.action);
			}
		}
	}

	private assignOwnerPath(owner: WorldObject, path: string, value: unknown, clear: boolean): void {
		const segments = path.split('.');
		if (segments.length === 0) return;
		let target: Record<string, unknown> = owner as unknown as Record<string, unknown>;
		for (let index = 0; index < segments.length - 1; index++) {
			const key = segments[index]!;
			let next = target[key];
			if (!next || typeof next !== 'object') {
				next = {};
				target[key] = next as never;
			}
			target = next as Record<string, unknown>;
		}
		const finalKey = segments[segments.length - 1]!;
		if (clear) {
			if (Array.isArray(target)) {
				delete (target as unknown as Record<string, unknown>)[finalKey];
			} else {
				delete target[finalKey];
			}
			return;
		}
		if (value && typeof value === 'object') {
			target[finalKey] = deep_clone(value as Record<string, unknown>);
			return;
		}
		target[finalKey] = value as never;
	}

	private resolveIntentPlayerInput(component: InputIntentComponent, owner: WorldObject): PlayerInput {
		const explicitIndex = component.playerIndex ?? 0;
		const fallback = (owner as { player_index?: number }).player_index ?? 0;
		const resolved = explicitIndex > 0 ? explicitIndex : fallback;
		// Require a positive player index; avoids silently reading player 0 (unset) and keeps intent bindings deterministic.
		if (resolved <= 0) {
			throw new Error(`[InputActionEffectSystem] Unable to resolve player index for object '${owner.id ?? '<unknown>'}'.`);
		}
		return $.input.getPlayerInput(resolved);
	}

	private resolveProgramKey(component: InputActionEffectComponent, owner: WorldObject): string {
		if (component.program_id) return component.program_id;
		return `inline:${owner.id}`;
	}

	private describeInlineProgram(component: InputActionEffectComponent): string {
		let ownerId: string;
		if (component.parent.id) {
			ownerId = component.parent.id;
		} else {
			ownerId = '<unattached>';
		}
		let componentId: string;
		if (component.id) {
			componentId = component.id;
		} else if (component.id_local) {
			componentId = component.id_local;
		} else {
			componentId = component.constructor.name;
		}
		return `inline:${ownerId}:${componentId}`;
	}

	private isEligibleObject(obj: WorldObject): boolean {
		if (obj.dispose_flag) return false;
		if (obj.active === false) return false;
		if (!obj.tick_enabled) return false;
		return true;
	}

	private evaluateProgram(
		program: CompiledProgram,
		env: BindingExecutionEnv,
		programKey: string,
	): void {
		const { input } = env;
		const bindings = program.bindings;
		for (let i = 0; i < bindings.length; i++) {
			const binding = bindings[i]!;
			if (!binding.predicate(env)) continue;

			const bindingKey = this.makeBindingKey(env.ownerId, programKey, env.playerIndex, binding, i);
			const armed = this.bindingLatch.get(bindingKey) === true;
			if (armed) this.frameLatchTouched.add(bindingKey);

			const pressMatched = binding.press ? binding.press(input) : false;
			const holdMatched = binding.hold ? binding.hold(input) : false;
			const releaseMatched = binding.release ? binding.release(input) : false;
			const customEdges = binding.customEdges;
			if (!armed && !pressMatched && !holdMatched && !releaseMatched && customEdges.length === 0) continue;

			const scratch = this.ensureScratch(customEdges.length);
			for (let j = 0; j < customEdges.length; j++) {
				scratch[j] = customEdges[j]!.match(input);
			}

			let matched = false;
			const runEffect = (effect: EffectExecutor): boolean => {
				if (!effect) return false;
				effect(env);
				return true;
			};

			if (pressMatched) {
				matched = true;
				if (binding.pressEffect) {
					if (runEffect(binding.pressEffect)) {
						this.bindingLatch.set(bindingKey, true);
						this.frameLatchTouched.add(bindingKey);
					}
				} else {
					this.bindingLatch.set(bindingKey, true);
					this.frameLatchTouched.add(bindingKey);
				}
			}
			if (holdMatched) {
				matched = true;
				if (binding.holdEffect) runEffect(binding.holdEffect);
				this.bindingLatch.set(bindingKey, true);
				this.frameLatchTouched.add(bindingKey);
			}
			if (releaseMatched && armed) {
				if (binding.releaseEffect && runEffect(binding.releaseEffect)) {
					matched = true;
				} else if (binding.releaseEffect === undefined) {
					matched = true;
				}
				this.bindingLatch.delete(bindingKey);
			}

			for (let j = 0; j < customEdges.length; j++) {
				if (!scratch[j]) continue;
				const effect = customEdges[j]!.effect;
				if (effect) {
					if (runEffect(effect)) matched = true;
				} else {
					matched = true;
				}
			}

			if (matched && program.evalMode === 'first') return;
		}
	}

	private makeBindingKey(ownerId: string, programKey: string, playerIndex: number, binding: CompiledBinding, index: number): string {
		const name = binding.name ?? `#${index}`;
		return `${ownerId}|${programKey}|p${playerIndex}|${name}|${index}`;
	}

	private ensureScratch(size: number): boolean[] {
		if (this.customMatchScratch.length < size) {
			const previousLength = this.customMatchScratch.length;
			this.customMatchScratch.length = size;
			for (let i = previousLength; i < size; i++) this.customMatchScratch[i] = false;
		}
		return this.customMatchScratch;
	}

	private resolveCompiledProgram(component: InputActionEffectComponent): CompiledProgram {
		if (component.program) {
			const program = component.program;
			if (!this.validatedInlinePrograms.has(program)) {
				const inlineId = this.describeInlineProgram(component);
				validateProgramEffects(program, inlineId);
				this.validatedInlinePrograms.add(program);
			}
			let compiled = this.inlineCompiled.get(program);
			if (!compiled) {
				compiled = compileProgram(program, pattern => this.parsePattern(pattern));
				this.inlineCompiled.set(program, compiled);
			}
			return compiled;
		}

		const programId = component.program_id;
		if (!programId) {
			const hostId = component.parent.id ?? component.id ?? '<unknown object>';
			throw new Error(`[InputActionEffectSystem] Component '${component.constructor.name}' on '${hostId}' is missing both an inline program and a programId.`);
		}

		let compiled = this.compiledById.get(programId);
		if (compiled) return compiled;

		const program = this.resolveProgramById(programId);

		compiled = compileProgram(program, pattern => this.parsePattern(pattern));
		this.compiledById.set(programId, compiled);
		return compiled;
	}

	private resolveProgramById(programId: string): InputActionEffectProgram {
		const cached = this.resolvedPrograms.get(programId);
		if (cached) return cached;
		if (this.missingProgramIds.has(programId)) {
			throw new Error(`[InputActionEffectSystem] Program '${programId}' is marked as missing.`);
		}

		const data = $.assets.data[tokenKeyFromId(programId) as keyof typeof $.assets.data];
		if (!isInputActionEffectProgram(data)) {
			this.missingProgramIds.add(programId);
			throw new Error(`[InputActionEffectSystem] Program '${programId}' not found or invalid.`);
		}

		this.resolvedPrograms.set(programId, data);
		return data;
	}

	private parsePattern(pattern: string): PatternPredicate {
		let predicate = this.patternCache.get(pattern);
		if (predicate) return predicate;
		predicate = (input: PlayerInput) => input.checkActionTriggered(pattern);
		this.patternCache.set(pattern, predicate);
		if (this.patternCache.size > this.patternCacheMax) {
			const iterator = this.patternCache.keys();
			const firstKey = iterator.next().value as string;
			if (firstKey !== undefined && firstKey !== pattern) this.patternCache.delete(firstKey);
		}
		return predicate;
	}
}

File: src/bmsx/res/bios/input_action_effect_system.lua
-- input_action_effect_system.lua
-- input intent + input action effect ecs system

local ecs<const> = require('ecs')
local action_effects<const> = require('action_effects')
local compiler<const> = require('input_action_effect_compiler')
local dsl<const> = require('input_action_effect_dsl')
local scratchbatch<const> = require('scratchbatch')
local world_instance<const> = require('world').instance
local inputintentcomponent<const> = 'inputintentcomponent'
local inputactioneffectcomponent<const> = 'inputactioneffectcomponent'
local actioneffectcomponentid<const> = 'actioneffectcomponent'
local assigned_value_edges<const> = { ['hold'] = true, ['press'] = true }
local active_scope<const> = { scope = 'active' }

local asset_programs_validated = false

local run_effect<const> = function(effect, env)
	if not effect then
		return false
	end
	effect(env)
	return true
end

local validate_primary_assets_on_boot<const> = function()
	if asset_programs_validated then
		return
	end
	for id, value in pairs(assets.data) do
		if dsl.is_input_action_effect_program(value) then
			compiler.validate_program_effects(value, id)
		end
	end
	asset_programs_validated = true
end

local inputactioneffectsystem<const> = {}
inputactioneffectsystem.__index = inputactioneffectsystem
setmetatable(inputactioneffectsystem, { __index = ecs.ecsystem })

function inputactioneffectsystem.new(priority)
	local self<const> = setmetatable(ecs.ecsystem.new(ecs.tickgroup.input, priority or 0), inputactioneffectsystem)
	self.compiled_by_id = {}
	self.inline_compiled = setmetatable({}, { __mode = 'k' })
	self.validated_inline = setmetatable({}, { __mode = 'k' })
	self.resolved_programs = {}
	self.missing_program_ids = {}
	self.pattern_cache = {}
	self.pattern_cache_max = 256
	self.custom_match_scratch = scratchbatch.new()
	self.runtime_by_component = setmetatable({}, { __mode = 'k' })
	self.frame_serial = 0
	self.__ecs_id = 'inputactioneffectsystem'
	validate_primary_assets_on_boot()
	return self
end

function inputactioneffectsystem:update()
	self.frame_serial = self.frame_serial + 1
	self:process_input_intents()
	self:process_input_action_programs()
end

function inputactioneffectsystem:process_input_intents()
	for obj, component in world_instance:objects_with_components(inputintentcomponent, active_scope) do
		if not (obj.active) then
			goto continue
		end
		if not component.bindings or #component.bindings == 0 then
			goto continue
		end
		local player_index<const> = self:resolve_intent_player_index(component, obj)
		for i = 1, #component.bindings do
			self:evaluate_intent_binding(obj, player_index, component.bindings[i])
		end
		::continue::
	end
end

function inputactioneffectsystem:process_input_action_programs()
	for obj, component in world_instance:objects_with_components(inputactioneffectcomponent, active_scope) do
		if not (obj.active) then
			goto continue
		end
		local program<const> = self:resolve_compiled_program(component)
		local program_key<const> = self:resolve_program_key(component, obj)

		local player_index<const> = obj.player_index or 1
		local effects<const> = obj:get_component(actioneffectcomponentid)
		if (not effects) and program.uses_effect_triggers then
			error('[inputactioneffectsystem] program "' .. program_key .. '" triggers effects but object "' .. obj.id .. '" has no actioneffectcomponent.')
		end

		local owner_id<const> = effects and effects.parent.id or obj.id
		local component_runtime<const> = self:resolve_component_runtime(component)
		local env<const> = component_runtime.env
		env.owner = obj
		env.owner_id = owner_id
		env.player_index = player_index
		env.effects = effects

		self:evaluate_program(program, env, program_key, component_runtime)
		local queued_commands<const> = env.queued_commands
		for i = 1, #queued_commands do
			local command<const> = queued_commands[i]
			obj:dispatch_command(command.event, command.payload)
			queued_commands[i] = nil
		end
		local queued<const> = env.queued_events
		for i = 1, #queued do
			local evt<const> = queued[i]
			obj:emit_gameplay_fact(evt)
			queued[i] = nil
		end
		::continue::
	end
end

function inputactioneffectsystem:evaluate_intent_binding(owner, player_index, binding)
	local action<const> = binding.action
	if not action then
		return
	end
	local state<const> = $.get_action_state(player_index, action)
	if state.justpressed and binding.press then
		self:run_intent_assignments(owner, player_index, binding, 'press', binding.press)
	end
	if state.pressed and binding.hold then
		self:run_intent_assignments(owner, player_index, binding, 'hold', binding.hold)
	end
	if state.justreleased and binding.release then
		self:run_intent_assignments(owner, player_index, binding, 'release', binding.release)
	end
end

function inputactioneffectsystem:run_intent_assignments(owner, player_index, binding, edge, spec)
	local assignments
	if type(spec) ~= 'table' or spec.path then
		assignments = { spec }
	else
		assignments = spec
	end
	for i = 1, #assignments do
		local assignment<const> = assignments[i]
		local path<const> = assignment.path
		local should_clear<const> = assignment.clear or (assignment.value == nil and edge == 'release')
			local resolved_value<const> = should_clear and nil or (assignment.value == nil and assigned_value_edges[edge] or assignment.value)
		self:assign_owner_path(owner, path, resolved_value, should_clear)
		if (assignment.consume) then
			consume_action(player_index, binding.action)
		end
	end
end

function inputactioneffectsystem:assign_owner_path(owner, path, value, clear)
	local segments<const> = {}
	for part in string.gmatch(path, '[^%.]+') do
		segments[#segments + 1] = part
	end
	local target = owner
	for i = 1, #segments - 1 do
		local key<const> = segments[i]
		local next_table = target[key]
		if type(next_table) ~= 'table' then
			next_table = {}
			target[key] = next_table
		end
		target = next_table
	end
	local final_key<const> = segments[#segments]
	if clear then
		target[final_key] = nil
		return
	end
	target[final_key] = value
end

function inputactioneffectsystem:resolve_intent_player_index(component, owner)
	local resolved<const> = component.player_index or owner.player_index
	if not resolved then
		error('[inputactioneffectsystem] unable to resolve player index for object "' .. (owner.id or '<unknown>') .. '".')
	end
	return resolved
end

function inputactioneffectsystem:resolve_program_key(component, owner)
	if component.program_id then
		return component.program_id
	end
	return 'inline:' .. owner.id
end

function inputactioneffectsystem:describe_inline_program(component)
	local owner_id<const> = component.parent and component.parent.id or '<unattached>'
	local component_id<const> = component.id or component.id_local or component.type_name or 'component'
	return 'inline:' .. owner_id .. ':' .. component_id
end

function inputactioneffectsystem:resolve_component_runtime(component)
	local component_runtime = self.runtime_by_component[component]
	if component_runtime then
		return component_runtime
	end
	local queued_commands<const> = {}
	local queued_events<const> = {}
	component_runtime = {
		binding_latch = {},
		binding_touched = {},
		binding_count = 0,
		last_frame = 0,
		queued_commands = queued_commands,
		queued_events = queued_events,
		env = {
			queued_commands = queued_commands,
			queued_events = queued_events,
		},
	}
	self.runtime_by_component[component] = component_runtime
	return component_runtime
end

function inputactioneffectsystem:reset_component_runtime(component_runtime, binding_count)
	local latch<const> = component_runtime.binding_latch
	local touched<const> = component_runtime.binding_touched
	local clear_count = component_runtime.binding_count
	if clear_count < binding_count then
		clear_count = binding_count
	end
	for i = 1, clear_count do
		latch[i] = false
		touched[i] = 0
	end
	component_runtime.binding_count = binding_count
end

function inputactioneffectsystem:prepare_component_runtime(component_runtime, program, program_key, env)
	local binding_count<const> = #program.bindings
	if component_runtime.last_frame ~= self.frame_serial - 1
		or component_runtime.program ~= program
		or component_runtime.program_key ~= program_key
		or component_runtime.owner_id ~= env.owner_id
		or component_runtime.player_index ~= env.player_index
		or component_runtime.binding_count ~= binding_count then
		self:reset_component_runtime(component_runtime, binding_count)
	end
	component_runtime.last_frame = self.frame_serial
	component_runtime.program = program
	component_runtime.program_key = program_key
	component_runtime.owner_id = env.owner_id
	component_runtime.player_index = env.player_index
	component_runtime.binding_count = binding_count
end

function inputactioneffectsystem:evaluate_program(program, env, program_key, component_runtime)
	self:prepare_component_runtime(component_runtime, program, program_key, env)
	local bindings<const> = program.bindings
	local frame<const> = self.frame_serial
	local latch<const> = component_runtime.binding_latch
	local touched<const> = component_runtime.binding_touched
	for i = 1, #bindings do
		local binding<const> = bindings[i]
		if not binding.predicate(env) then
			goto continue
		end

		local armed<const> = latch[i]
		if armed then
			touched[i] = frame
		end

		local press_matched<const> = binding.press and binding.press(env) or false
		local hold_matched<const> = binding.hold and binding.hold(env) or false
		local release_matched<const> = binding.release and binding.release(env) or false
		local custom_edges<const> = binding.custom_edges
		if not armed and not press_matched and not hold_matched and not release_matched and #custom_edges == 0 then
			goto continue
		end

		local scratch<const> = self.custom_match_scratch:reserve(#custom_edges, false)
		for j = 1, #custom_edges do
			scratch[j] = custom_edges[j].match(env)
		end

		local matched

		if press_matched then
			matched = true
			if binding.press_effect then
				if run_effect(binding.press_effect, env) then
					latch[i] = true
					touched[i] = frame
				end
			else
				latch[i] = true
				touched[i] = frame
			end
		end
		if hold_matched then
			matched = true
			if binding.hold_effect then
				run_effect(binding.hold_effect, env)
			end
			latch[i] = true
			touched[i] = frame
		end
		if release_matched and armed then
			if binding.release_effect and run_effect(binding.release_effect, env) then
				matched = true
			elseif binding.release_effect == nil then
				matched = true
			end
			latch[i] = false
			touched[i] = 0
		end

		for j = 1, #custom_edges do
			if scratch[j] then
				local effect<const> = custom_edges[j].effect
				if effect then
					if run_effect(effect, env) then
						matched = true
					end
				else
					matched = true
				end
			end
		end

		if matched and program.eval_mode == 'first' then
			break
		end

		::continue::
	end
	for i = 1, component_runtime.binding_count do
		if latch[i] and touched[i] ~= frame then
			latch[i] = false
		end
	end
end

function inputactioneffectsystem:resolve_compiled_program(component)
	if component.program then
		local program<const> = component.program
		if not self.validated_inline[program] then
			compiler.validate_program_effects(program, self:describe_inline_program(component))
			self.validated_inline[program] = true
		end
		local compiled = self.inline_compiled[program]
		if not compiled then
			compiled = compiler.compile_program(program, function(pattern)
				return self:parse_pattern(pattern)
			end)
			self.inline_compiled[program] = compiled
		end
		return compiled
	end

	local program_id<const> = component.program_id
	if not program_id then
		error('[inputactioneffectsystem] component on "' .. (component.parent and component.parent.id or '<unknown>') .. '" is missing program_id.')
	end

	local compiled = self.compiled_by_id[program_id]
	if compiled then
		return compiled
	end

	local program<const> = self:resolve_program_by_id(program_id)
	compiled = compiler.compile_program(program, function(pattern)
		return self:parse_pattern(pattern)
	end)
	self.compiled_by_id[program_id] = compiled
	return compiled
end

function inputactioneffectsystem:resolve_program_by_id(program_id)
	if self.resolved_programs[program_id] then
		return self.resolved_programs[program_id]
	end
	if self.missing_program_ids[program_id] then
		error('[inputactioneffectsystem] program "' .. program_id .. '" is marked as missing.')
	end
	local data<const> = assets.data[program_id]
	if not dsl.is_input_action_effect_program(data) then
		self.missing_program_ids[program_id] = true
		error('[inputactioneffectsystem] program "' .. program_id .. '" not found or invalid.')
	end
	self.resolved_programs[program_id] = data
	return data
end

function inputactioneffectsystem:parse_pattern(pattern)
	local predicate = self.pattern_cache[pattern]
	if predicate then
		return predicate
	end
	predicate = function(env)
		return action_triggered(pattern, env.player_index)
	end
	self.pattern_cache[pattern] = predicate
	if self.pattern_cache_max and (function()
		local count = 0
		for _ in pairs(self.pattern_cache) do
			count = count + 1
		end
		return count
	end)() > self.pattern_cache_max then
		for key in pairs(self.pattern_cache) do
			if key ~= pattern then
				self.pattern_cache[key] = nil
				break
			end
		end
	end
	return predicate
end

return {
	inputactioneffectsystem = inputactioneffectsystem,
}
